[ENH] Forecasting metrics dispatch of `__call__` to `evaluate` vs `evaluate_by_index`

benHeid opened the pull request:
<!--
Welcome to sktime, and thanks for contributing!
Please have a look at our contribution guide:
https://www.sktime.net/en/latest/get_involved/contributing.html
-->

#### Reference Issues/PRs
Closes #6813 

#### What does this implement/fix? Explain your changes.
Adds a class parameter to dispatch between `evaluate` and `evaluate_by_index`, if `__call__` is called.


#### Did you add any tests for the change?


#### Any other comments?
<!--
We value all user contributions, no matter how small or complex they are. If you have any questions, feel free to post
in the dev-chat channel on the sktime discord https://discord.com/invite/54ACzaFsn7. If we are slow to review (>3 working days), likewise feel free to ping us on discord. Thank you for your understanding during the review process.
-->

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

fkiraly reviewed (changes requested):
Great, I already forgot about this but this is a neat addition!

Only point, I think it should be shorter for convenience of use. `by_index`?

benHeid commented:
@fkiraly thanks for the review. I added a new test and removed the get_test_params. I have added them since a lot of the metrics do only have one test set, leading to failing tests...

fkiraly reviewed (changes requested):
the `get_test_params` tests only fail if you have non-reserved params and there are not at least two settings.

If you add `by_index` to the `reserved_params` tag in `BaseForecastingErrorMetric`, most of the tests should hopefully pass?

It should be a reserved parameter, because it is shared by all metrics and now part of the base API.

fkiraly commented on sktime/performance_metrics/forecasting/tests/test_all_metrics_forecasting.py:
```diff
@@ -186,3 +186,25 @@ def test_metric_weights(self, estimator_instance):
         # res_wt = wt_metr(y_true, y_pred)
         # assert np.allclose(res_wt, metric_obj(y_true, y_pred, sample_weight=wts))
+
+    def test_evaluate_by_index(self, estimator_instance):
```
hm, looking at the tests, I would have added it at the end of `test_metric_output_by_instance`, as follows:

* clone the metric used above and `set_params` the `by_index`, or equivalent copy/set logic
* execute the `__call__`
* check that it is identical with the return (using `pandas` equals)

This works because we checked all conventions on `evaluate_by_index` above already, and we now only need to additionally check that we get the same from the call.

fkiraly commented:
current failures likely due to incomplete renames, or some metrics missing the new param.

benHeid commented:
> current failures likely due to incomplete renames, or some metrics missing the new param.

I have overseen some classes where I needed to add the `by_index` param. The remaining failures seems to be related to TSFresh.

fkiraly reviewed (approved):
Yes, the merge from `main` should have fixed the `tsfresh` related failure.

I've updated the docstring to make it clearer what the parameter does, hope this is fine.
